File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    //TreeNode* test = new TreeNode(2, new TreeNode(4, new TreeNode(6), nullptr), new TreeNode(5));
    
    bool nodeEquality(TreeNode* root, TreeNode* subRoot) {
        if (subRoot == nullptr && root == nullptr) { return true; }
        if (!subRoot || !root) { return false; }
        if (root->val != subRoot->val) { return false; }
        return nodeEquality(root->left, subRoot->left) && nodeEquality(root->right, subRoot->right);
    }

    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (!root) { return false; }
        if (nodeEquality(root, subRoot)) { return true; }
        return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
    }
};

File: Data Structures & Algorithms/same-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

def sameTree(p, q):
    if not p and not q:
        return True
    if not p or not q:
        return False

    if p.val != q.val:
        return False
    if not sameTree(p.left, q.left):
        return False
    if not sameTree(p.right, q.right):
        return False
    return True

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        return sameTree(p, q)
